File: combinatorics/resource/LCHP/problem_LTDT_code/bai_14/bai14.py
import heapq
from typing import List, Tuple, Dict, Optional
import sys

class SimpleDijkstra:
    """Dijkstra cho đồ thị đơn"""
    
    def __init__(self, num_vertices: int):
        self.num_vertices = num_vertices
        self.adj_list = [[] for _ in range(num_vertices)]
        self.reset()
    
    def reset(self):
        self.dist = [float('inf')] * self.num_vertices
        self.parent = [-1] * self.num_vertices
        self.visited = [False] * self.num_vertices
    
    def add_edge(self, u: int, v: int, weight: int):
        """Thêm cạnh vào đồ thị vô hướng"""
        self.adj_list[u].append((v, weight))
        self.adj_list[v].append((u, weight))
        print(f"Thêm cạnh {u} - {v} với trọng số {weight}")
    
    def dijkstra_with_heap(self, source: int):
        """Dijkstra sử dụng heap"""
        self.reset()
        
        # Heap: (distance, vertex)
        heap = [(0, source)]
        self.dist[source] = 0
        
        print(f"\nDijkstra với heap từ đỉnh {source}:")
        
        while heap:
            current_dist, u = heapq.heappop(heap)
            
            if self.visited[u]:
                continue
                
            self.visited[u] = True
            print(f"Xử lý đỉnh {u} với khoảng cách {current_dist}")
            
            # Duyệt các đỉnh kề
            for v, weight in self.adj_list[u]:
                if not self.visited[v]:
                    new_dist = self.dist[u] + weight
                    
                    if new_dist < self.dist[v]:
                        self.dist[v] = new_dist
                        self.parent[v] = u
                        heapq.heappush(heap, (new_dist, v))
                        
                        print(f"  Cập nhật dist[{v}] = {new_dist} (qua đỉnh {u})")
    
    def dijkstra_basic(self, source: int):
        """Dijkstra cơ bản với mảng"""
        self.reset()
        
        self.dist[source] = 0
        
        print(f"\nDijkstra cơ bản từ đỉnh {source}:")
        
        for count in range(self.num_vertices):
            # Tìm đỉnh chưa xử lý có khoảng cách nhỏ nhất
            u = self._find_min_distance()
            if u == -1:
                break
                
            self.visited[u] = True
            print(f"Bước {count + 1}: Chọn đỉnh {u} với khoảng cách {self.dist[u]}")
            
            # Cập nhật khoảng cách đến các đỉnh kề
            for v, weight in self.adj_list[u]:
                if (not self.visited[v] and 
                    self.dist[u] != float('inf') and 
                    self.dist[u] + weight < self.dist[v]):
                    
                    self.dist[v] = self.dist[u] + weight
                    self.parent[v] = u
                    
                    print(f"  Cập nhật dist[{v}] = {self.dist[v]}")
    
    def _find_min_distance(self) -> int:
        """Tìm đỉnh chưa xử lý có khoảng cách nhỏ nhất"""
        min_dist = float('inf')
        min_vertex = -1
        
        for v in range(self.num_vertices):
            if not self.visited[v] and self.dist[v] < min_dist:
                min_dist = self.dist[v]
                min_vertex = v
                
        return min_vertex
    
    def print_results(self, source: int):
        """In kết quả Dijkstra"""
        print(f"\nKết quả Dijkstra từ đỉnh {source}:")
        print(f"{'Đỉnh':>6} {'Khoảng cách':>12} {'Đường đi':>15}")
        print("-" * 35)
        
        for i in range(self.num_vertices):
            print(f"{i:>6}", end="")
            
            if self.dist[i] == float('inf'):
                print(f"{'∞':>12} {'Không có':>15}")
            else:
                print(f"{self.dist[i]:>12}", end="")
                path_str = self._get_path_string(source, i)
                print(f"{path_str:>15}")
        print()
    
    def _get_path_string(self, source: int, dest: int) -> str:
        """Lấy chuỗi biểu diễn đường đi"""
        if dest == source:
            return str(dest)
        
        if self.parent[dest] == -1:
            return "Không có đường"
        
        path = []
        current = dest
        
        while current != -1:
            path.append(current)
            current = self.parent[current]
        
        path.reverse()
        return "→".join(map(str, path))
    
    def print_graph(self):
        """In đồ thị"""
        print("Đồ thị đơn có trọng số:")
        for i in range(self.num_vertices):
            print(f"Đỉnh {i}: ", end="")
            for v, weight in self.adj_list[i]:
                print(f"({v},{weight}) ", end="")
            print()
        print()
    
    def shortest_path(self, source: int, dest: int) -> Tuple[int, List[int]]:
        """Tìm đường đi ngắn nhất giữa hai đỉnh"""
        self.dijkstra_with_heap(source)
        
        if self.dist[dest] == float('inf'):
            return (-1, [])
        
        path = []
        current = dest
        
        while current != -1:
            path.append(current)
            current = self.parent[current]
        
        path.reverse()
        return (self.dist[dest], path)
    
    def analyze_graph(self):
        """Phân tích đồ thị"""
        print("Phân tích đồ thị:")
        
        # Kiểm tra tính liên thông
        self.dijkstra_with_heap(0)
        is_connected = all(dist != float('inf') for dist in self.dist)
        
        print(f"Đồ thị liên thông: {'Có' if is_connected else 'Không'}")
        
        # Tìm cặp đỉnh có khoảng cách lớn nhất
        max_dist = 0
        farthest_pair = (-1, -1)
        
        for i in range(self.num_vertices):
            self.dijkstra_with_heap(i)
            for j in range(self.num_vertices):
                if self.dist[j] != float('inf') and self.dist[j] > max_dist:
                    max_dist = self.dist[j]
                    farthest_pair = (i, j)
        
        if farthest_pair[0] != -1:
            print(f"Cặp đỉnh xa nhất: {farthest_pair[0]} và {farthest_pair[1]} "
                  f"(khoảng cách: {max_dist})")
        print()

def main():
    """Demo Dijkstra cho đồ thị đơn"""
    print("=== DIJKSTRA CHO ĐỒ THỊ ĐƠN ===\n")
    
    graph = SimpleDijkstra(5)
    
    # Tạo đồ thị mẫu
    graph.add_edge(0, 1, 4)
    graph.add_edge(0, 2, 2)
    graph.add_edge(1, 2, 1)
    graph.add_edge(1, 3, 5)
    graph.add_edge(2, 3, 8)
    graph.add_edge(2, 4, 10)
    graph.add_edge(3, 4, 2)
    
    print()
    graph.print_graph()
    
    # Dijkstra với heap
    graph.dijkstra_with_heap(0)
    graph.print_results(0)
    
    # Dijkstra cơ bản
    graph.dijkstra_basic(0)
    graph.print_results(0)
    
    # Tìm đường đi cụ thể
    dist, path = graph.shortest_path(0, 4)
    print("Đường đi ngắn nhất từ 0 đến 4:")
    print(f"Độ dài: {dist}")
    print(f"Đường đi: {'→'.join(map(str, path))}\n")
    
    # Phân tích đồ thị
    graph.analyze_graph()

if __name__ == "__main__":
    main()

File: combinatorics/resource/LCHP/problem_LTDT_code/bai_14/bai14.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <climits>
#include <algorithm>
#include <iomanip>

using namespace std;

struct Edge {
    int to;
    int weight;
    
    Edge(int t, int w) : to(t), weight(w) {}
};

class SimpleDijkstra {
private:
    int numVertices;
    vector<vector<Edge>> adjList;
    vector<int> dist;
    vector<int> parent;
    vector<bool> visited;
    
public:
    SimpleDijkstra(int n) : numVertices(n) {
        adjList.resize(n);
        reset();
    }
    
    void reset() {
        dist.assign(numVertices, INT_MAX);
        parent.assign(numVertices, -1);
        visited.assign(numVertices, false);
    }
    
    void addEdge(int u, int v, int weight) {
        adjList[u].push_back(Edge(v, weight));
        adjList[v].push_back(Edge(u, weight)); // Đồ thị vô hướng
        cout << "Thêm cạnh " << u << " - " << v << " với trọng số " << weight << endl;
    }
    
    // Dijkstra sử dụng priority queue
    void dijkstraWithPQ(int source) {
        reset();
        
        // Priority queue (min-heap): pair<distance, vertex>
        priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> pq;
        
        dist[source] = 0;
        pq.push({0, source});
        
        cout << "\nDijkstra với Priority Queue từ đỉnh " << source << ":" << endl;
        
        while (!pq.empty()) {
            int u = pq.top().second;
            int d = pq.top().first;
            pq.pop();
            
            if (visited[u]) continue;
            
            visited[u] = true;
            cout << "Xử lý đỉnh " << u << " với khoảng cách " << d << endl;
            
            // Duyệt các đỉnh kề
            for (const Edge& edge : adjList[u]) {
                int v = edge.to;
                int weight = edge.weight;
                
                if (!visited[v] && dist[u] + weight < dist[v]) {
                    dist[v] = dist[u] + weight;
                    parent[v] = u;
                    pq.push({dist[v], v});
                    
                    cout << "  Cập nhật dist[" << v << "] = " << dist[v] 
                         << " (qua đỉnh " << u << ")" << endl;
                }
            }
        }
    }
    
    // Dijkstra cơ bản với mảng
    void dijkstraBasic(int source) {
        reset();
        
        dist[source] = 0;
        
        cout << "\nDijkstra cơ bản từ đỉnh " << source << ":" << endl;
        
        for (int count = 0; count < numVertices - 1; count++) {
            // Tìm đỉnh chưa xử lý có khoảng cách nhỏ nhất
            int u = findMinDistance();
            if (u == -1) break;
            
            visited[u] = true;
            cout << "Bước " << count + 1 << ": Chọn đỉnh " << u 
                 << " với khoảng cách " << dist[u] << endl;
            
            // Cập nhật khoảng cách đến các đỉnh kề
            for (const Edge& edge : adjList[u]) {
                int v = edge.to;
                int weight = edge.weight;
                
                if (!visited[v] && dist[u] != INT_MAX && 
                    dist[u] + weight < dist[v]) {
                    dist[v] = dist[u] + weight;
                    parent[v] = u;
                    
                    cout << "  Cập nhật dist[" << v << "] = " << dist[v] << endl;
                }
            }
        }
    }
    
private:
    int findMinDistance() {
        int minDist = INT_MAX;
        int minVertex = -1;
        
        for (int v = 0; v < numVertices; v++) {
            if (!visited[v] && dist[v] < minDist) {
                minDist = dist[v];
                minVertex = v;
            }
        }
        
        return minVertex;
    }
    
public:
    void printResults(int source) {
        cout << "\nKết quả Dijkstra từ đỉnh " << source << ":" << endl;
        cout << setw(6) << "Đỉnh" << setw(12) << "Khoảng cách" 
             << setw(15) << "Đường đi" << endl;
        cout << string(35, '-') << endl;
        
        for (int i = 0; i < numVertices; i++) {
            cout << setw(6) << i;
            
            if (dist[i] == INT_MAX) {
                cout << setw(12) << "∞" << setw(15) << "Không có" << endl;
            } else {
                cout << setw(12) << dist[i] << setw(15);
                printPath(source, i);
                cout << endl;
            }
        }
        cout << endl;
    }
    
    void printPath(int source, int dest) {
        if (dest == source) {
            cout << dest;
            return;
        }
        
        if (parent[dest] == -1) {
            cout << "Không có đường";
            return;
        }
        
        vector<int> path;
        int current = dest;
        
        while (current != -1) {
            path.push_back(current);
            current = parent[current];
        }
        
        reverse(path.begin(), path.end());
        
        for (int i = 0; i < path.size(); i++) {
            cout << path[i];
            if (i < path.size() - 1) cout << "→";
        }
    }
    
    void printGraph() {
        cout << "Đồ thị đơn có trọng số:" << endl;
        for (int i = 0; i < numVertices; i++) {
            cout << "Đỉnh " << i << ": ";
            for (const Edge& edge : adjList[i]) {
                cout << "(" << edge.to << "," << edge.weight << ") ";
            }
            cout << endl;
        }
        cout << endl;
    }
    
    // Tìm đường đi ngắn nhất giữa hai đỉnh cụ thể
    pair<int, vector<int>> shortestPath(int source, int dest) {
        dijkstraWithPQ(source);
        
        if (dist[dest] == INT_MAX) {
            return {-1, {}};
        }
        
        vector<int> path;
        int current = dest;
        
        while (current != -1) {
            path.push_back(current);
            current = parent[current];
        }
        
        reverse(path.begin(), path.end());
        return {dist[dest], path};
    }
    
    // Kiểm tra tính chất của đồ thị
    void analyzeGraph() {
        cout << "Phân tích đồ thị:" << endl;
        
        // Kiểm tra tính liên thông
        bool isConnected = true;
        dijkstraWithPQ(0);
        
        for (int i = 0; i < numVertices; i++) {
            if (dist[i] == INT_MAX) {
                isConnected = false;
                break;
            }
        }
        
        cout << "Đồ thị liên thông: " << (isConnected ? "Có" : "Không") << endl;
        
        // Tìm cặp đỉnh có khoảng cách lớn nhất
        int maxDist = 0;
        pair<int, int> farthestPair = {-1, -1};
        
        for (int i = 0; i < numVertices; i++) {
            dijkstraWithPQ(i);
            for (int j = 0; j < numVertices; j++) {
                if (dist[j] != INT_MAX && dist[j] > maxDist) {
                    maxDist = dist[j];
                    farthestPair = {i, j};
                }
            }
        }
        
        if (farthestPair.first != -1) {
            cout << "Cặp đỉnh xa nhất: " << farthestPair.first 
                 << " và " << farthestPair.second 
                 << " (khoảng cách: " << maxDist << ")" << endl;
        }
        
        cout << endl;
    }
};

// Demo cho đồ thị đơn
int main() {
    cout << "=== DIJKSTRA CHO ĐỒ THỊ ĐƠN ===" << endl << endl;
    
    SimpleDijkstra graph(5);
    
    // Tạo đồ thị mẫu
    graph.addEdge(0, 1, 4);
    graph.addEdge(0, 2, 2);
    graph.addEdge(1, 2, 1);
    graph.addEdge(1, 3, 5);
    graph.addEdge(2, 3, 8);
    graph.addEdge(2, 4, 10);
    graph.addEdge(3, 4, 2);
    
    cout << endl;
    graph.printGraph();
    
    // Dijkstra với priority queue
    graph.dijkstraWithPQ(0);
    graph.printResults(0);
    
    // Dijkstra cơ bản
    graph.dijkstraBasic(0);
    graph.printResults(0);
    
    // Tìm đường đi cụ thể
    auto path = graph.shortestPath(0, 4);
    cout << "Đường đi ngắn nhất từ 0 đến 4:" << endl;
    cout << "Độ dài: " << path.first << endl;
    cout << "Đường đi: ";
    for (int i = 0; i < path.second.size(); i++) {
        cout << path.second[i];
        if (i < path.second.size() - 1) cout << "→";
    }
    cout << endl << endl;
    
    // Phân tích đồ thị
    graph.analyzeGraph();
    
    return 0;
}